### PR TITLE
Vulkan: Multi-window support

### PR DESCRIPTION
**Pull request by thatcosmonaut**
This patch uses a dictionary of window handles to swapchains to blit the faux-backbuffer to the appropriate swapchain to handle multiple windows on the same device.

Some outstanding concerns: Now that we aren't using a one swapchain assumption, the initialization flow feels slightly awkward. This is related to the special cases for `deviceWindowHandle` in the swapchain functions. We might want to consider initializing with a dummy surface similar to the `PrepareWindowAttributes` flow and then handling windows/surfaces/swapchains as always-bundled units, but I'm not sure if this would also be awkward.

### CODE REVIEW

**Comment by thatcosmonaut** [comment]
Tweaked the initialization flow - now every window, including the default window, has its own surface and swapchain via the window hash lookup.

**Comment by flibitijibibo** [comment]
The hash table seems excessive for something that won't have that much data... would it make sense to just have an array of swapchain info and assign them to windows via SDL_SetWindowData?

**Comment by thatcosmonaut** [comment]
It's a 1-to-1 association so that would probably work... let me think about this.

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6087]:
```suggestion
	uint32_t swapchainImageIndex
) {
```

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6255]: Looks like this TODO is obsolete now?

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6659]:
```suggestion
	/* Each swapchain must have its own surface. */
```

**Review by flibitijibibo — CHANGES_REQUESTED** [review]
Minor details aside from the hashmap work.

**Comment by thatcosmonaut** [comment]
Alright, it uses `SDL_SetWindowData` and `SDL_GetWindowData` instead of a hash now.

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6161]:
```suggestion
		swapchainData = (VulkanSwapchainData*) SDL_GetWindowData(windowHandle, WINDOW_SWAPCHAIN_DATA);
```

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6171]:
```suggestion
			swapchainData = (VulkanSwapchainData*) SDL_GetWindowData(windowHandle, WINDOW_SWAPCHAIN_DATA);
```

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6527]:
```suggestion
	swapchainData = (VulkanSwapchainData*) SDL_GetWindowData(windowHandle, WINDOW_SWAPCHAIN_DATA);
```

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6689]:
```suggestion
	swapchainData = (VulkanSwapchainData*) SDL_GetWindowData(windowHandle, WINDOW_SWAPCHAIN_DATA);
```

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:164]: Might need something more obvious for anyone inspecting via debugger:

```suggestion
#define WINDOW_SWAPCHAIN_DATA "FNA3D_VulkanSwapchain"
```

This will help differentiate it from, say, a D3D11 swapchain pair whenever that gets supported.

**Review by flibitijibibo — APPROVED** [review]
Few more tiny things, otherwise this looks fine. We'll want to get this in the hands of Guard Crush and YoYo for live testing. If we think D3D11 will be as straightforward as this we can hold off until both are ready to look at.

**Review by flibitijibibo — APPROVED** [review]
Should be okay. Aside from the aforementioned: @kg, I believe this affects the SRGB extension so this may be worth checking even with a single window.

**Comment by thatcosmonaut** [comment]
Noticed a validation error when testing Flotilla, might want to hold off on merging until I take care of that.

**Comment by thatcosmonaut** [comment]
Validation error fixed, I wasn't querying the surface for swapchain support.

**Comment by thatcosmonaut** [comment]
Murder Miners trace lit up some backbuffer resize issues, fixing those.

**Comment by thatcosmonaut** [comment]
We've got a problem - I've resolved some swapchain management issues but it seems like `ResetBackbuffer` is triggering before the window is done resizing, so the surface query parameters are outdated and we try to generate a swapchain that is larger than the window. This is on Fedora 35 Wayland. Happening on master as well, if you run the Murder Miners trace you'll see a Warn about swapchain sizing and the game will only draw at the original size before window resize.

**Comment by flibitijibibo** [comment]
Probably related:

https://github.com/libsdl-org/SDL/issues/4595

https://github.com/libsdl-org/SDL/issues/4563

**Comment by thatcosmonaut** [comment]
Alright, looks like this isn't our renderer's fault at least. I'll re-add the extents check warn.

**Comment by thatcosmonaut** [comment]
Confirmed that SDL_VIDEODRIVER=wayland does not exhibit this behavior, so this is an XWayland issue for sure.

**Comment by thatcosmonaut** [comment]
Ugh it was my fault after all, had a bad realloc in there. Not really sure why the swapchain size is wrong on master but it's fixed on both XWayland and Wayland with this most recent commit.

**Comment by thatcosmonaut** [comment]
Gonna run some traces again. This patchset has ended up being a pretty large improvement to Vulkan swapchain management in general, I'm thinking we should maybe merge this in after testing and I can rebase my buffer sync rewrite on this once it's in.

**Comment by thatcosmonaut** [comment]
Traces look fine, will wait for another review.

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6639]: Remaining TODO. In particular, the 0,0 extent is actually really important for surface suspension on Windows and Wayland, so we'll need to reincorporate this where it's been removed.

Other than that this is all fine.

**Comment by thatcosmonaut** [comment]
Re-added zero size surface handling and tightened up the swapchain handling in SwapBuffers a bit more for failure cases.

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:41]: Might need a less generic name, though I doubt it will collide with anything...

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:236]: Can probably bring back the comma if only to shrink the diff size

```suggestion
	CREATE_SWAPCHAIN_SURFACE_ZERO,
```

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:1027]: Cleaning this up as best as I can...
```suggestion
	/* Window surface */
	VkSurfaceKHR surface;
	VkSurfaceFormatKHR surfaceFormat;
	void *windowHandle;

	/* Swapchain for window surface */
	VkSwapchainKHR swapchain;
	VkFormat swapchainFormat;
	VkComponentMapping swapchainSwizzle;
	VkPresentModeKHR presentMode;

	/* Swapchain images */
	VkExtent2D extent;
	VkImage *images;
	VkImageView *views;
	VulkanResourceAccessType *resourceAccessTypes;
	uint32_t imageCount;
```

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:2083]: This can warn, will error later
```suggestion
		FNA3D_LogWarn("This surface does not support presenting!");
```

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:2074]: I think we check this somewhere else... does one make the other redundant by chance?

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:2295]: Nit: Totally agree with the new capitalization but we should probably do that separately to make this diff smaller.

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6012]: Nit: Same thing with the capitalization...

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6168]: This might need a cast... surprised it didn't need one before, actually.

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6415]: Doesn't `!validSwapchainExists` imply `!acquireSuccess`?

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6627]: LogError might throw, so these frees may need to be above that.

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:6701]: Same thing as the previous comment, LogError should be after frees.

File: src/FNA3D_Driver_Vulkan.c
Comment (flibitijibibo) [inline src/FNA3D_Driver_Vulkan.c:12458]: This can be 1 by default, 99.99% of uses will be single-window.

**Review by flibitijibibo — COMMENTED** [review]
One final pass to be as thorough as possible, these are all pretty isolated so once these are done I'll merge it in right away. Definitely an improvement even for single-window setups!

File: src/FNA3D_Driver_Vulkan.c
Comment (thatcosmonaut) [reply src/FNA3D_Driver_Vulkan.c:2074]: This gets checked when we do the dummy surface in PrepareWindowAttributes, but we have to check it for every surface we create or Vulkan validation yells at us.

File: src/FNA3D_Driver_Vulkan.c
Comment (thatcosmonaut) [reply src/FNA3D_Driver_Vulkan.c:2300]: Won't this make the spelling inconsistent in the patch though?

**Comment by thatcosmonaut** [comment]
Put in all the suggestions except for the capitalization thing. If you're fine with the capitalization being inconsistent in the patch then I'll add it in.

**Comment by thatcosmonaut** [comment]
Found one more small thing when testing Flotilla, fixing now...

**Comment by thatcosmonaut** [comment]
Alright, took care of that. Looks like Flotilla is losing track of a few VkBuffers on exit, is that expected?

**Comment by flibitijibibo** [comment]
Like FEZ it's pretty leaky, so nothing to worry about there.

**Pull request merged**
